lib/charms/hpc_libs/v0/slurm_ops.py

```diff
@@ -197,9 +148,183 @@ def config_name(self) -> str:
+            target = key.rsplit(".", maxsplit=1)[-1]
+            configs[target] = config
+
+        return configs
+
+    @abstractmethod
+    def get(self, key: Optional[str] = None) -> Any:
+        """Get specific configuration value for Slurm component."""
+        ...
+
+    @abstractmethod
+    def set(self, config: Mapping[str, Any]) -> None:
+        """Set configuration for Slurm component."""
+        ...
+
+    @abstractmethod
+    def unset(self, *keys: str) -> None:
+        """Unset configuration for Slurm component."""
+        ...
+
+
+class MungekeyManager(ABC):
+    """Control the munge key."""
+
+    def get(self) -> str:
+        """Get the current munge key.
+
+        Returns:
+            The current munge key as a base64-encoded string.
+        """
+        ...
+
+    def set(self, key: str) -> None:
+        """Set a new munge key.
+
+        Args:
+            key: A new, base64-encoded munge key.
+        """
+        ...
+
+    def generate(self) -> None:
+        """Generate a new, cryptographically secure munge key."""
+        ...
+
+
+class SlurmPackageManager(ABC):
+    """Manager to control the installation, creation and configuration of Slurm-related services."""
+
+    @abstractmethod
+    def install(self) -> None:
+        """Install Slurm."""
+        ...
+
+    @abstractmethod
+    def version(self) -> str:
+        """Get the current version of Slurm installed on the system."""
+        ...
+
+    @abstractmethod
+    def service_for(self, service: ServiceType) -> ServiceManager:
+        """Get a `ServiceManager` for the specified `service`."""
+        ...
+
+    @abstractmethod
+    def configuration_manager_for(self, service: ServiceType) -> ConfigurationManager:
+        """Get a `ConfigurationManager` for the specified `service`."""
+        ...
+
+    @abstractmethod
+    def mungekey_manager(self) -> MungekeyManager:
+        """Get the `MungekeyManager` for munge."""
+        ...
+
+
+class MungeManager:
+    """Manage `munged` service operations."""
+
+    def __init__(self, ops_manager: SlurmPackageManager) -> None:
+        self.service = ops_manager.service_for(ServiceType.MUNGED)
+        self.config = ops_manager.configuration_manager_for(ServiceType.MUNGED)
+        self.key = ops_manager.mungekey_manager()
+
+
+class PrometheusExporterManager:
+    """Manage `slurm-prometheus-exporter` service operations."""
+
+    def __init__(self, ops_manager: SlurmPackageManager) -> None:
+        self.service = ops_manager.service_for(ServiceType.PROMETHEUS_EXPORTER)
+
+
+class SlurmManagerBase:
+    """Base manager for Slurm services."""
+
+    def __init__(self, service: ServiceType, ops_manager: SlurmPackageManager) -> None:
+        self.service = ops_manager.service_for(service)
+        self.config = ops_manager.configuration_manager_for(service)
+        self.munge = MungeManager(ops_manager)
+        self.exporter = PrometheusExporterManager(ops_manager)
+        self.ops = ops_manager
```

Comment:
I'm having a thought here :monocle_face: 

What if we used a boolean flag here to indicate which operations manager we're using rather than needing to directly pass a package manager object? We know that we're only going to support deb and snap, so it shouldn't be too complicated to implement:

```python3
def __init__(self, service: ServiceType, snap: bool = False) -> None:
    ops_manager = SnapPackageManager() if snap else DebPackageManager()
    self.service = ops_manager.service_manager
    self.config = ops_manager.config_manager
    self.munge = ops_manager.munge_manager
    self.exporter = ops_manager.exporter_manager
    self.install = ops_manager.install
    self.version = ops_manager.version
```

This way we can have the default behavior be to use the Debian package, change snap to True when we're ready to switch over, and we can bind the operations manager methods and properties to where we need them to be. We also don't have to expose the methods for retrieving the format-specific operations managers. We also don't need to have the `ops` attribute then to expose the `install` and `version` methods and can just call them the way that we did previously.

Let me know your thoughts on this!

Reply:
Good idea! That's both simpler for us, and it could also be extended later if we do have to support an additional package manager.